### ISSUE TITLE
Fail to catch SocketException while connecting.

### ISSUE THREAD

**Issue by butlermatt** [dart-lang/sdk#27440]
I'm having difficulties creating a replication for this issue in a simple case. However the long and short of it is:

I have code similar to this:

```dart
Socket sock;
Future connect() async {
  try {
    sock = await Socket.connect(ipaddress, port);
    print('Connected??');
  } catch (e) {
    print('Caught error: $e');
    sock?.destroy();
    return new Future.delayed(new Duration(seconds: 5), () {
      connect();
    });
  }

  print('Connected!!');
}
```

There is more code but that's the general gist of it. If I have a connection timed out when trying to connect I catch that error without a problem. If I get an error can't assign requested address I can catch that no problem. But if it looks like the connection is about to work, and I receive a `SocketException: OS Error: Connection reset by peer, errno = 54, address = 192.168.168.32, port = 63130` then that error is not caught. It does not print the `connected??` line so I know that's where it's failing, but I can't explain why I can't catch the error.

As stated I can catch other errors without any issue. I can also catch the Connection reset by peer when the connection itself is established (handled in the `sock.listen(_handleMessage, onError: () { ..});`) but if I receive connection reset by peer before the connection is fully established (eg before the await statement completes and the socket is assigned to the variable) I can't catch it.

**Comment by whesse** [comment]
This sounds like a quite possible error in our implementation - the code that gets the errors from the C++ io pool threads, and connects them to the right Dart object or future is quite complex, and we could be missing this case somehow.

**Comment by johnmccutchan** [comment]
@zanderso

**Comment by kaendfinger** [comment]
This is possibly the cause of bad reconnection behavior issues we are seeing in our IoT platform. We have two servers connecting to each other over a fast but unstable network, and if the network goes bad for a minute, we try to reconnect and it fails very quickly, we seem to see this issue, so our reconnection logic is unable to continue correctly.

**Comment by kaendfinger** [comment]
This is still hitting us really hard, is there any way to have this fixed?

**Comment by zanderso** [comment]
https://chromiumcodereview.appspot.com/2426413006/

I haven't been able to reproduce this, but this CL fixes some code that looked very suspicious. It may or may not fix the problem. Do you have a reliable way to test this?

**Comment by kaendfinger** [comment]
Sure, I'll test it ASAP.

**Comment by zanderso** [comment]
@kaendfinger @butlermatt Have either of you had a chance to try out the fix?

**Comment by kaendfinger** [comment]
yes, so it looks like the problem is not solved, I'll get a more stable test case and provide more information on the environments on where we see it happen.

**Comment by SandPod** [comment]
Any updates on this?

Did a post here with some more information: https://github.com/dart-lang/sdk/issues/31558#issuecomment-2516595095

**Comment by abitofevrything** [comment]
Not sure if this is entirely related, but I was running into a similar issue where I would be unable to catch `SocketException`s from a Socket. In my case, this was after the connection was successfully established, and I had a listener with an `onError` callback. The `onError` callback successfully fired, but I was still seeing an uncaught error.

Turns out the socket implementation actually reports errors _twice_: once as an error event on the stream, and then by completing the `Socket.done` future with an error. Having an `onError` handler for the stream does not prevent the future from reporting an uncaught error. Not sure if the reverse is true (i.e handling errors on the future but not the stream).

Simply adding `socket.done.ignore()` to my code fixed it. But obviously that isn't possible here (as the error occurrs before the `Socket` instance is accessible by the user), and it definitely feels like a hack that shouldn't be needed. But it could be why this issue appears to handle errors and still get uncaught errors reported.
